src/java.base/share/classes/java/util/stream/GathererOp.java

```diff
@@ -0,0 +1,755 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package java.util.stream;
+
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.Spliterator;
+import java.util.concurrent.CountedCompleter;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.BinaryOperator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.function.ToDoubleFunction;
+import java.util.function.ToIntFunction;
+import java.util.function.ToLongFunction;
+import java.util.stream.Gatherer.Integrator;
+
+/**
+ * Runtime machinery for evaluating Gatherers under different modes.
+ * The performance-critical code below contains some more complicated encodings:
+ * therefore, make sure to run benchmarks to verify changes to prevent regressions.
+ *
+ * @since 22
+ */
+final class GathererOp<T, A, R> extends ReferencePipeline<T, R> {
+    @SuppressWarnings("unchecked")
+    static <P_IN, P_OUT extends T, T, A, R> Stream<R> of(
+            ReferencePipeline<P_IN, P_OUT> upstream,
+            Gatherer<T, A, R> gatherer) {
+        // When attaching a gather-operation onto another gather-operation,
+        // we can fuse them into one
+        if (upstream.getClass() == GathererOp.class) {
+            return new GathererOp<>(
+                    ((GathererOp<P_IN, Object, P_OUT>) upstream).gatherer.andThen(gatherer),
+                    (GathererOp<?, ?, P_IN>) upstream);
+        } else {
+            return new GathererOp<>(
+                    (ReferencePipeline<?, T>) upstream,
+                    gatherer);
+        }
+    }
+
+    /*
+     * GathererOp.NodeBuilder is a lazy accumulator of elements with O(1)
+     * `append`, and O(8) `join` (concat).
+     *
+     * First `append` inflates a growable Builder, the O(8) for `join` is
+     * because we prefer to delegate to `append` for small concatenations to
+     * avoid excessive indirections (unbalanced Concat-trees) when joining many
+     * NodeBuilders together.
+     */
+    final static class NodeBuilder<X> implements Consumer<X> {
+        private final static int LINEAR_APPEND_MAX = 8; // TODO revisit
+        final static class Builder<X> extends SpinedBuffer<X> implements Node<X> {
```

Comment:
Let's use blessed modifiers order:
```suggestion
    static final class NodeBuilder<X> implements Consumer<X> {
        private static final int LINEAR_APPEND_MAX = 8; // TODO revisit
        static final class Builder<X> extends SpinedBuffer<X> implements Node<X> {
```